File: linked-list/linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        fast = slow = head
        while fast:
            fast = fast.next
            if fast:
                slow = slow.next
                fast = fast.next
                if fast and fast == slow:
                    return True
        return False
       

File: linked-list/linked-list-cycle.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {Boolean}
def hasCycle(head)
    return false if head.nil?

    slow = fast = head
    while fast
        fast = fast.next
        return true if fast == slow
        if fast
            fast = fast.next
            slow = slow.next
        end
    end
    false
end